### ISSUE TITLE
discrete output

### ISSUE THREAD

**Issue by Zoe-Queen** [GuyLor/direct_vae#1]
In your work, discrete variables are implemented in latent variables. May I ask whether this method is feasible for discrete variables to be directly used as network output?

**Comment by GuyLor** [comment]
Yes. However, in this case maybe you should consider other directions.

If you have a differentiable loss function, you can train with continuous layers and discretize the predictions during inference, as commonly done in classifiers.

If your loss/reward requires discrete inputs, "direct loss optimization" without Gumbel perturbation can be used (although, in most cases, the perturbation won't hurt). Alternatively, you can use the "straight through" method, that is, using a Softmax layer with discretization in the forward pass and relaxed in the backward pass. In order to sample from the network output before calculating the loss function, Gumbel perturbations are needed.
